### ISSUE TITLE
Keypad entry for exact scores in the Enter Score dialog

### ISSUE THREAD

**Issue by mkieres** [Progrunning/BoardGamesCompanion#94]
> **Spec** — synthesised from a design session. Records the decision in [ADR-0005](../blob/main/docs/adr/0005-keypad-entry-for-exact-scores.md) (`docs/adr/0005-keypad-entry-for-exact-scores.md`).

## Problem Statement

Recording a score in the Enter Score dialog builds the total by **accumulating partial scores** — each tap of a `+1/+5/+10/+50` tile, or a spin of the circular dial, appends one increment and the running total is their sum. This makes large totals painful: a player who scores 1000 taps `+50` twenty times, and money/economic games routinely go higher and land on exact odd figures (1350, 247) that no round increment reaches. Users have told us entering very high scores is cumbersome.

## Solution

Let a player **type an exact score on a keypad** instead of only accumulating it. Tapping a keypad tile in the instant-score panel swaps the dialog's dial region for an in-app numeric keypad; the player types any integer up to six digits and commits it. The typed number is added (or subtracted, following the `+/−` toggle) as a single partial score, so it behaves like every other entry — it shows in the history strip and Undo pops it. Entering a value into an empty score lands an exact total in one action, which is what money games need.

To make room without redesigning the dialog, the `+50` tile is removed and the keypad tile takes its slot: the instant-score row becomes `operation · {1, 5, 10, keypad}`, keeping the current footprint. The circular dial is left untouched.

## User Stories

1. As a player recording a money-game score, I want to type 1350 directly, so that I don't tap `+50` twenty-seven times.
2. As a player, I want to open a keypad from the Enter Score dialog, so that entering a large number is fast.
3. As a player, I want the keypad to appear in place of the dial within the same dialog, so that I stay in one focused surface and don't lose my running score.
4. As a player, I want the score and the `+/−` toggle to stay visible while I type, so that I can see whether I'm adding or subtracting the number.
5. As a player, I want my typed number added to the current score, so that I can top up a score I've already started building.
6. As a player, I want to switch the `+/−` toggle to subtract and then type, so that I can record a large penalty or a negative swing.
7. As a player, I want my typed value to appear in the `(+…, +…)` history strip like any other partial score, so that I can review what made up the total.
8. As a player, I want to Undo a typed value, so that a mistyped number is as easy to remove as a mis-tapped tile.
9. As a player, I want to type into an empty score and get exactly that total, so that I can enter a final count directly.
10. As a player, I want the keypad to accept only whole numbers, so that the score matches how it's shown everywhere else in the app.
11. As a player, I want to correct a digit while typing (backspace), so that I don't have to cancel and start over.
12. As a player, I want to cancel out of the keypad without committing, so that opening it by accident costs me nothing.
13. As a player, I want to commit the typed value and return to the normal dialog view, so that I can keep adjusting with the tiles or dial afterwards.
14. As a player, I want the keypad to stop me entering absurdly long numbers, so that the score stays readable in the display and history.
15. As a player who scored in the old way, I want the `1/5/10` tiles and the dial to keep working exactly as before, so that nothing I rely on breaks.
16. As a player, I want the Done button to still assume 0 when I enter nothing, so that the existing quick-exit behaviour is preserved.

## Implementation Decisions

- **Modules touched:** the Enter Score dialog widget (`enter_score_dialog.dart`) and its view model (`EnterScoreViewModel`). No model/schema changes — a typed value flows through the same `scoreGameResult.points` path as any partial score.
- **Keypad logic lives in the view model, not the widget.** The view model owns the in-progress typed digits and exposes actions to append a digit, backspace, cancel, and commit. Committing produces a single partial score (negated when `operation == subtract`) and routes through the existing `updateScore` path so history, `canUndo`, and the sum-of-partials invariant are unchanged. The widget is a thin renderer over this state.
- **Surface: in-dialog swap.** Opening the keypad replaces the circular-dial region inside the existing dialog card; the score display and `+/−` operation toggle stay pinned above it. No bottom sheet, no system keyboard, no keyboard-inset handling. The card already sizes to its content.
- **Custom keypad, not the OS keyboard.** Built from the existing tile widgets (`_InstantScoreTile` / `ElevatedContainer`) for visual consistency and full control over the input rules below.
- **Input rules:** integer-only; capped at six digits (≤ 999999); sign comes from the existing `+/−` toggle, so there is no minus key on the pad. Matches the `toStringAsFixed(0)` display used across the app.
- **The `+50` tile is removed** and the keypad tile takes its place: `operation · {1, 5, 10, keypad}`. This keeps the instant-score row's footprint identical (252px fixed against a 308px content width at the 340px minimum), where a fifth tile would overflow on small phones.
- **History rendering:** a committed typed value is an ordinary partial score with no special treatment — e.g. `+1350` in the `(+…, +…)` strip.
- Full rationale and rejected alternatives (#94's `100/1000` tiles, a `×10` multiplier, per-game increment sets, the system keyboard) are recorded in **ADR-0005**.

## Testing Decisions

- **What a good test asserts:** external behaviour of the view model — given a starting score, an operation, and a sequence of keypad actions, what is the resulting score, partial-score history, and `canUndo` state. Not widget internals or render structure.
- **Module under test:** `EnterScoreViewModel`, unit-tested in isolation. This is the single, highest seam — pushing the keypad's digit-accumulation and sign logic into the view model means the whole feature is verifiable here without pumping widgets.
- **Prior art:** `test/view_models/edit_playthrough_view_model_test.dart` — same MobX-store-in-isolation style with `mocktail` mocks and hand-built `PlayerScore` fixtures. Follow its setup.
- **Cases to cover:** typing then committing adds a partial equal to the number; commit with `subtract` adds a negative partial; typing into an empty score yields exactly that total; six-digit cap rejects a seventh digit; backspace edits in-progress digits; cancel commits nothing and leaves score/history untouched; Undo pops a committed typed value; existing `1/5/10`, dial, and `scoreZero`-on-Done paths still behave.
- A widget test asserting the dial↔keypad swap and that the score/toggle stay visible is welcome but secondary; the correctness contract lives at the view-model seam.

## Out of Scope

- **"Top 5 by victories, not points"** — the original second half of this ticket. It is a separate statistics concern, unrelated to score entry, and should move to its own issue rather than ride along here.
- **Removing or redesigning the circular dial.** Deliberately gated on a Discord usage/sentiment poll; not decided here (see ADR-0005 consequences).
- **Per-game scoring styles / configurable increment sets.** No per-game score type exists today; the keypad makes it unnecessary for this problem.
- **Decimal scores.** Nothing in the app displays decimals; not introduced here.

## Further Notes

- The `+50` removal is an accepted, pragmatic trade — the dial covers mid-range jumps and the keypad covers large ones, so `50` was the most redundant tile. Flagged here because a reviewer who scored in 50s will notice it's gone.
- A Discord poll (usage + sentiment on the dial) is planned to inform the separate dial-redesign decision.

**Comment by mkieres** [comment]
+1
